cumulusci/tasks/salesforce/metadata_etl/base.py

```diff
@@ -0,0 +1,262 @@
+    """Base class for Metadata ETL tasks that generate new metadata
+    and deploy it into the org."""
+
+    deploy = True
+
+    def _generate_package_xml(self, deploy):
+        generator = PackageXmlGenerator(str(self.deploy_dir), self.api_version)
+        return generator()
+
+    def _transform(self):
+        self._synthesize()
+
+    def _synthesize(self):
+        # Create metadata in self.deploy_dir
+        pass
+
+
+class BaseMetadataTransformTask(BaseMetadataETLTask):
+    """Base class for Metadata ETL tasks that extract metadata,
+    transform it, and deploy it back into the org."""
+
+    retrieve = True
+    deploy = True
+
+    def _get_entities(self):
+        return {}
+
+    def _get_types_package_xml(self):
+        base = """    <types>
+{members}
+        <name>{name}</name>
+    </types>
+"""
+        types = ""
+        for entity, api_names in self._get_entities().items():
+            members = "\n".join(
+                f"        <members>{api_name}</members>" for api_name in api_names
+            )
+            types += base.format(members=members, name=entity)
+
+        return types
+
+    def _get_package_xml_content(self, deploy):
+        return f"""<?xml version="1.0" encoding="UTF-8"?>
+<Package xmlns="http://soap.sforce.com/2006/04/metadata">
+{self._get_types_package_xml()}
+    <version>{self.api_version}</version>
+</Package>
+"""
+
+    def _transform(self):
+        pass
+
+
+class MetadataSingleEntityTransformTask(BaseMetadataTransformTask):
+    """Base class for a Metadata ETL task that affects one or more
+    instances of a specific metadata entity."""
+
+    entity = None
+
+    task_options = {
+        "api_names": {"description": "List of API names of entities to affect"},
+        **BaseMetadataETLTask.task_options,
+    }
+
+    def _init_options(self, kwargs):
+        super()._init_options(kwargs)
+
+        self.api_names = list(
+            map(
+                self._namespace_injector,
+                process_list_arg(self.options.get("api_names", [])),
+            )
+        )
+
+    def _get_entities(self):
+        return {self.entity: self.api_names or ["*"]}
+
+    def _transform_entity(self, metadata, api_name):
+        return metadata
+
+    def _transform(self):
+        # call _transform_entity once per retrieved entity
+        # if the entity is an XML file, provide a parsed version
+        # and write the returned metadata into the deploy directory
+
+        parser = PackageXmlGenerator(
+            None, self.api_version
+        )  # We'll use it for its metadata_map
+        entity_configurations = list(
+            filter(
+                lambda entry: any(
+                    [
+                        subentry["type"] == self.entity
+                        for subentry in parser.metadata_map[entry]
+                    ]
+                ),
+                parser.metadata_map,
+            )
+        )
```

Comment:
entity_configurations = [entry for entry in parser.metadata_map if any( ... )]

But maybe the metadata_map should be wrapped by a class that provides a way to look up by type, instead of implementing that here.